src/main/java/io/rtr/conduit/amqp/impl/AMQPTransport.java

```diff
@@ -3,25 +3,68 @@
 import com.rabbitmq.client.AlreadyClosedException;
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.MetricsCollector;
+import com.rabbitmq.client.ShutdownListener;
+import com.rabbitmq.client.ShutdownSignalException;
 import io.rtr.conduit.amqp.AMQPConsumerCallback;
 import io.rtr.conduit.amqp.AMQPMessageBundle;
 import io.rtr.conduit.amqp.AbstractAMQPTransport;
 import io.rtr.conduit.amqp.transport.TransportConnectionProperties;
 import io.rtr.conduit.amqp.transport.TransportListenProperties;
 import io.rtr.conduit.amqp.transport.TransportMessageBundle;
 import io.rtr.conduit.amqp.transport.TransportPublishProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.io.IOException;
 import java.time.Duration;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.TimeoutException;
 
 public class AMQPTransport extends AbstractAMQPTransport {
+
+    class DynamicQueueCleanupShutdownListener implements ShutdownListener {
+        //@VisibleForTesting
+        CompletableFuture<Void> queueCleanupJob;
+
+        @Override
+        public void shutdownCompleted(ShutdownSignalException e) {
+            // Shutdown handlers run directly in the AMQP Connection's worker loop.
+            // As such, trying to initiate new AMQP commands from within it creates a deadlock.
+            queueCleanupJob = CompletableFuture.runAsync(()-> {
+                if (dynamicQueue != null) {
+                    deleteQueueOnSeparateChannel(dynamicQueue);
+                    LOGGER.debug("Deleted dynamic queue '{}'.", dynamicQueue);
+                }
+                if (dynamicPoisonQueueInUse) {
+                    deleteQueueOnSeparateChannel(POISON + "." + dynamicQueue);
+                    LOGGER.debug("Deleted poison message queue for dynamic queue '{}'.", dynamicQueue);
+                }
+            });
+        }
+
+        private void deleteQueueOnSeparateChannel(String queue) {
+            try (Channel cleanupChannel = conn.createChannel()) {
+                try {
+                    cleanupChannel.queueDelete(queue);
```

Comment:
`queueDelete()` throws `IOException`, maybe narrow the exception type.  Also, do you want to re-throw this exception so that we end up propagating it like we're allowing other exceptions to?  If we also fail to create the channel (meaning we don't end up deleting as well), we don't log any errors, so, this could all probably logged in the outer catch right before the re-throw, with no inner try-catch.

Reply:
Since `createChannel` can also throw `RuntimeException`, e.g. `ConduitConnectionStateException`, we may want to catch/log that as well

Reply:
The shutdown listener is executed on the connection's worker thread used internally by Rabbit, so I'm not sure how a rethrown exception would be surfaced in a way consuming code could handle.

But yes we only need on e try block here, will update

Reply:
Further investigation shows internally in the rabbit client, any exceptions in shutdown handlers are caught in a block like this in a `ShutdownNotifierComponent` class:
```
            try {
                l.shutdownCompleted(sse);
            } catch (Exception e) {
            // FIXME: proper logging
            }
```
¯\_(ツ)_/¯